src/modules/base/ResourcesIndex.ts

```diff
@@ -0,0 +1,232 @@
+import { Logger } from "../../standard/log/Logger.ts";
+import { SearchIndex } from "../../standard/handlers/SearchIndex.ts";
+import { StatusCode } from "../../standard/http/response/StatusCode.ts";
+
+// Imports > Modules
+import { ResourceClassesArray } from "./types/ResourceClassesArray.ts";
+
+type Index = {
+  resource: IResource;
+  path_patterns: PathPattern[];
+};
+
+type Input = { url: string };
+
+type PathPatternExecResult = {
+  pathname?: {
+    groups: Record<string, string | undefined>;
+  };
+};
+
+interface PathPattern {
+  pathname: string;
+  exec(input: string): PathPatternExecResult | null;
+}
+
+type PathPatternClass = ConstructorWithArgs<
+  PathPattern,
+  [{ pathname: string }]
+>;
+
+type SearchResult = {
+  resource: IResource;
+  path_params: Record<string, string | undefined>;
+};
+
+abstract class ResourcesIndex extends SearchIndex<
+  Promise<SearchResult | null>
+> {
+  #cached_search_results: Record<string, SearchResult | null> = {};
+  #logger: Logger = ConsoleLogger.create("ResourcesIndex", Level.Off);
+  protected index: Index[] = [];
+  protected resources: ResourceClassesArray = [];
+  protected PathPatternClass: PathPatternClass;
+
+  constructor(
+    PathPatternClass: PathPatternClass,
+    resources: ResourceClassesArray,
+  ) {
+    super();
+    this.resources = resources ?? [];
+    this.PathPatternClass = PathPatternClass;
+    this.buildIndex(this.resources);
+  }
+
+  public handle(request: Input): Promise<SearchResult | null> {
+    return Promise
+      .resolve()
+      .then(() => this.#logger.debug(`Input received`))
+      .then(() => this.#validateRequest(request))
+      .then(() => this.search(request))
+      .then((result) => super.nextHandler({ request, result }));
+  }
+
+  protected buildIndex(resources: ResourceClassesArray): void {
+    this.#logger.debug(`Building resources index`);
+
+    for (const ResourceClass of resources) {
+      if (Array.isArray(ResourceClass)) {
+        this.buildIndex(ResourceClass);
+        continue;
+      }
+
+      const pathPatterns: PathPattern[] = [];
+
+      const resource = new ResourceClass();
+      resource.paths.forEach((path: string) => {
+        // Add "{/}?" to match possible trailing slashes too. For example, this
+        // means the following paths point to the same resource:
+        //
+        //   - /coffee
+        //   - /coffee/
+        //
+        pathPatterns.push(
+          new this.PathPatternClass({ pathname: path + "{/}?" }),
+        );
+
+        this.#logger.debug(`Added resource/pathname mapping: {}`, {
+          name: resource.constructor.name,
+          path,
+        });
+      });
+
+      this.index.push({
+        resource,
+        path_patterns: pathPatterns,
+      });
+    }
+  }
+
+  protected search(request: { url: string }): Promise<SearchResult | null> {
+    const fullyQualifiedUrl = request.url;
+
+    this.#logger.debug(`Searching for resource - url: {}`, fullyQualifiedUrl);
```

Comment:
Another instance of logging that should not be the default behaviour.

Reply:
done in https://github.com/crookse/fork-drash/pull/1/commits/60e2c967262636b6b409618c3cd5b051e01312db and https://github.com/drashland/drash/pull/698/commits/daea29f90c5f3a8d415c8be08df050612c14b7af (same commits as above)